File: web/prayer-engine.test.js
const test = require("node:test");
const assert = require("node:assert/strict");
const { evaluate } = require("./prayer-engine.js");

function day(date, offset = 0) {
  const shift = (text) => {
    const [hour, minute] = text.split(":").map(Number);
    const total = hour * 60 + minute + offset;
    return `${String(Math.floor(total / 60)).padStart(2, "0")}:${String(total % 60).padStart(2, "0")}`;
  };
  return {
    date,
    fajr: { adhan: shift("05:04"), iqamah: shift("05:30") },
    dhuhr: { adhan: shift("13:08"), iqamah: shift("13:30") },
    asr: { adhan: shift("16:30"), iqamah: shift("18:15") },
    maghrib: { adhan: shift("19:31"), iqamah: shift("19:33") },
    isha: { adhan: shift("20:51"), iqamah: shift("21:15") },
  };
}

const payload = {
  prayer_days: [day("2026-09-13"), day("2026-09-14", 2)],
  content: [
    { id: "q1", date: "*", type: "QURAN", arabic: "آية", english: "Verse", reference: "2:153" },
    { id: "h1", date: "*", type: "HADITH", arabic: "حديث", english: "Hadith", reference: "1" },
  ],
};

const at = (hour, minute, second = 0) => ({ date: "2026-09-13", hour, minute, second });

test("main screen rolls a completed prayer to tomorrow", () => {
  const state = evaluate(at("10", "01"), payload);
  assert.equal(state.type, "main");
  assert.equal(state.rows[0].isTomorrow, true);
  assert.equal(state.rows[0].adhan, "05:06");
  assert.equal(state.rows[1].isNext, true);
});

test("shows Adhan for one minute", () => {
  const state = evaluate(at("13", "08", 30), payload);
  assert.deepEqual({ type: state.type, prayer: state.prayer, time: state.time }, { type: "adhan", prayer: "DHUHR", time: "13:08" });
});

test("shows the two-minute Iqamah countdown", () => {
  const state = evaluate(at("13", "28", 30), payload);
  assert.equal(state.type, "iqamah");
  assert.equal(state.prayer, "DHUHR");
  assert.equal(state.secondsRemaining, 90);
});

test("Adhan notice takes precedence at Maghrib", () => {
  assert.equal(evaluate(at("19", "31", 20), payload).type, "adhan");
  const countdown = evaluate(at("19", "32", 0), payload);
  assert.equal(countdown.type, "iqamah");
  assert.equal(countdown.secondsRemaining, 60);
});

test("rotates content for thirty seconds every five minutes", () => {
  assert.equal(evaluate(at("10", "05", 10), payload).type, "content");
  assert.equal(evaluate(at("10", "05", 35), payload).type, "main");
});

test("suppresses content for fifteen minutes after Iqamah", () => {
  assert.equal(evaluate(at("13", "35", 10), payload).type, "main");
  assert.equal(evaluate(at("13", "45", 10), payload).type, "content");
});

File: app/src/test/java/com/siea/prayerdisplay/domain/PrayerEngineTest.kt
package com.siea.prayerdisplay.domain

import com.siea.prayerdisplay.data.ContentEntity
import com.siea.prayerdisplay.data.PrayerDayEntity
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.ZoneId
import java.time.ZonedDateTime

class PrayerEngineTest {
    private val zone = ZoneId.of("Europe/London")
    private val days = listOf(day("2026-09-13", "05:00", "05:30"), day("2026-09-14", "05:05", "05:35"))
    private val content = listOf(
        ContentEntity("q1", "*", "QURAN", "Arabic", "English", "Reference", "")
    )

    @Test
    fun `adhan screen lasts for first minute after adhan`() {
        val state = evaluate("2026-09-13T05:00:30")
        assertTrue(state is DisplayState.Adhan)
        assertEquals(PrayerName.FAJR, (state as DisplayState.Adhan).prayer)
    }

    @Test
    fun `iqamah countdown begins two minutes before iqamah`() {
        val state = evaluate("2026-09-13T05:28:30")
        assertTrue(state is DisplayState.IqamahCountdown)
        assertEquals(PrayerName.FAJR, (state as DisplayState.IqamahCountdown).prayer)
        assertEquals(90, state.secondsRemaining)
    }

    @Test
    fun `completed prayer rolls both times to tomorrow at exact iqamah`() {
        val state = evaluate("2026-09-13T05:30:00") as DisplayState.Main
        val fajr = state.rows.first { it.name == PrayerName.FAJR }
        assertEquals("05:05", fajr.adhan)
        assertEquals("05:35", fajr.iqamah)
        assertTrue(fajr.isTomorrow)
    }

    @Test
    fun `daily content is suppressed for fifteen minutes after iqamah`() {
        assertTrue(evaluate("2026-09-13T05:35:05") is DisplayState.Main)
        assertTrue(evaluate("2026-09-13T05:40:05") is DisplayState.Main)
    }

    @Test
    fun `daily content resumes at end of quiet period`() {
        assertTrue(evaluate("2026-09-13T05:45:05") is DisplayState.DailyContent)
    }

    @Test
    fun `future prayers stay on current day`() {
        val state = evaluate("2026-09-13T05:30:00") as DisplayState.Main
        val dhuhr = state.rows.first { it.name == PrayerName.DHUHR }
        assertFalse(dhuhr.isTomorrow)
    }

    private fun evaluate(localDateTime: String): DisplayState = PrayerEngine.evaluate(
        ZonedDateTime.parse("${localDateTime}+01:00[Europe/London]"),
        days,
        content
    )

    private fun day(date: String, fajrAdhan: String, fajrIqamah: String) = PrayerDayEntity(
        date = date,
        fajrAdhan = fajrAdhan,
        fajrIqamah = fajrIqamah,
        dhuhrAdhan = "13:00",
        dhuhrIqamah = "13:30",
        asrAdhan = "17:00",
        asrIqamah = "17:15",
        maghribAdhan = "19:30",
        maghribIqamah = "19:35",
        ishaAdhan = "21:00",
        ishaIqamah = "21:15"
    )
}
